File: 1- Data Structures/1- Basic/4- Linked Lists/linked-list.js
class Node {
  constructor(data) {
      this.data = data;
      this.next = null;
  }
}

class LinkedList {
  constructor() {
      this.head = null;
  }

  // Adicionar um elemento no final da lista
  append(data) {
      let newNode = new Node(data);
      if (this.head === null) {
          this.head = newNode;
      } else {
          let current = this.head;
          while (current.next !== null) {
              current = current.next;
          }
          current.next = newNode;
      }
  }

  // Remover o primeiro elemento da lista
  removeFirst() {
      if (this.head !== null) {
          this.head = this.head.next;
      } else {
          console.log("A lista está vazia");
      }
  }

  // Exibir os elementos da lista
  display() {
      if (this.head === null) {
          console.log("A lista está vazia");
          return;
      }
      let current = this.head;
      while (current !== null) {
          console.log(current.data);
          current = current.next;
      }
  }
}

// Exemplo de utilização
let list = new LinkedList();
list.append(10);
list.append(20);
list.append(30);

console.log("Elementos da lista:");
list.display();  // Output: 10 20 30

console.log("Removendo o primeiro elemento");
list.removeFirst();
list.display();  // Output: 20 30

File: 1- Data Structures/1- Basic/4- Linked Lists/linked-list.cpp
#include <iostream>
using namespace std;

// Definição de um nó da lista
struct Node {
    int data;
    Node* next;
};

// Classe da lista encadeada
class LinkedList {
private:
    Node* head;

public:
    // Construtor
    LinkedList() {
        head = nullptr;
    }

    // Adicionar um elemento no final da lista
    void append(int data) {
        Node* newNode = new Node();
        newNode->data = data;
        newNode->next = nullptr;

        if (head == nullptr) {
            head = newNode;
        } else {
            Node* current = head;
            while (current->next != nullptr) {
                current = current->next;
            }
            current->next = newNode;
        }
    }

    // Remover o primeiro elemento da lista
    void removeFirst() {
        if (head != nullptr) {
            Node* temp = head;
            head = head->next;
            delete temp;
        } else {
            cout << "A lista está vazia" << endl;
        }
    }

    // Exibir os elementos da lista
    void display() {
        if (head == nullptr) {
            cout << "A lista está vazia" << endl;
            return;
        }
        Node* current = head;
        while (current != nullptr) {
            cout << current->data << " ";
            current = current->next;
        }
        cout << endl;
    }
};

// Exemplo de utilização
int main() {
    LinkedList list;
    list.append(10);
    list.append(20);
    list.append(30);

    cout << "Elementos da lista:" << endl;
    list.display();  // Output: 10 20 30

    cout << "Removendo o primeiro elemento" << endl;
    list.removeFirst();
    list.display();  // Output: 20 30

    return 0;
}
